syne_tune/stopping_criterions/automatic_termination_criterion.py

```diff
@@ -0,0 +1,159 @@
+    :param warm_up: Defines the minimum umber of data points before we start fitting the GP.
+    :param topq: We consider only the top q-% data points to fit the GP.
+    """
+
+    def __init__(
+        self,
+        config_space: Dict[str, Any],
+        metric: str,
+        threshold: float,
+        mode: str = "min",
+        beta: float = 1.0,
+        seed: int = None,
+        warm_up: int = 10,
+        topq: float = 0.5,
+    ):
+        assert mode in ["min", "max"]
+
+        self._mode = mode
+        self._warm_up = warm_up
+        self._seed = np.random.randint(2**32 - 1) if seed is None else seed
+        self._metric = metric
+        self._topq = topq
+        self._beta = beta
+        self._threshold = threshold
+        self._config_space = config_space
+        self._hp_ranges = make_hyperparameter_ranges(config_space)
+
+        if self._mode == "max":
+            self.multiplier = 1
+        else:
+            self.multiplier = -1
+
+    def _config_to_feature_matrix(self, configs: List[dict]) -> Tensor:
+        bounds = Tensor(self._hp_ranges.get_ndarray_bounds()).T
+        X = Tensor(self._hp_ranges.to_ndarray_matrix(configs))
+        return normalize(X, bounds)
+
+    def __call__(self, status: TuningStatus) -> bool:
+        """Return a boolean representing if the tuning has to stop."""
+
+        trials = status.trial_rows
+
+        if len(trials) == 0:
+            return False
+
+        evaluations = []
+        observations = []
+        for trial in trials.values():
+            if self._metric not in trial:
+                continue
+            y = self.multiplier * trial[self._metric]
+            evaluations.append(y)
+
+            observation = {}
+            for hp in self._config_space:
+                observation[hp] = trial[hp]
+            observations.append(observation)
+
+        if len(observations) < self._warm_up:
+            return False
+
+        observations = self._config_to_feature_matrix(observations).double()
+        evaluations = torch.tensor(evaluations).double()
+
+        noise_std = NOISE_LEVEL
+        evaluations += noise_std * torch.randn_like(evaluations)
+
+        n = len(observations)
+
+        if self._topq is not None:
+            topn = max(20, int(n * self._topq))
+            top_inds = torch.argsort(evaluations, descending=True, dim=0)[:topn]
+            evaluations = evaluations[top_inds]
+            observations = observations[top_inds]
+
+        if len(observations.shape) == 1:
+            observations = observations.reshape((-1, 1))
+        if len(evaluations.shape) == 1:
+            evaluations = evaluations.reshape((-1, 1))
+
+        gp = SingleTaskGP(observations, evaluations)
+        mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
+        fit_gpytorch_mll(mll, max_attempts=1)
+        acq = UpperConfidenceBound(
+            model=gp,
+            beta=self._beta,
+        )
+        _, max_ucb = optimize_acqf(
```

Comment:
We should try catch this one. In case cholesky fails (which will be frequent with 3 restarts), we should not stop.

Reply:
All of such stuff is already there. You can replicate all this by using BOTorch, or you can just use the internal code that is already there.